libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/api/MapboxManeuverApi.kt

```diff
@@ -0,0 +1,119 @@
+package com.mapbox.navigation.ui.maneuver.api
+
+import com.mapbox.api.directions.v5.models.BannerComponents
+import com.mapbox.api.directions.v5.models.BannerInstructions
+import com.mapbox.api.directions.v5.models.LegStep
+import com.mapbox.api.directions.v5.models.RouteLeg
+import com.mapbox.navigation.base.formatter.DistanceFormatter
+import com.mapbox.navigation.base.trip.model.RouteStepProgress
+import com.mapbox.navigation.ui.base.api.maneuver.ManeuverApi
+import com.mapbox.navigation.ui.base.api.maneuver.ManeuverCallback
+import com.mapbox.navigation.ui.base.api.maneuver.StepDistanceRemainingCallback
+import com.mapbox.navigation.ui.base.api.maneuver.UpcomingManeuversCallback
+import com.mapbox.navigation.ui.base.model.maneuver.Maneuver
+import com.mapbox.navigation.ui.base.model.maneuver.ManeuverState
+import com.mapbox.navigation.ui.base.model.maneuver.ManeuverState.CurrentManeuver
+import com.mapbox.navigation.ui.base.model.maneuver.ManeuverState.DistanceRemainingToFinishStep
+import com.mapbox.navigation.ui.maneuver.ManeuverAction
+import com.mapbox.navigation.ui.maneuver.ManeuverProcessor
+import com.mapbox.navigation.ui.maneuver.ManeuverResult
+import com.mapbox.navigation.utils.internal.JobControl
+import com.mapbox.navigation.utils.internal.ThreadController
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+
+/**
+ * Implementation of [ManeuverApi] allowing access to generate representation of
+ * maneuvers based on [BannerInstructions]
+ * @property distanceFormatter DistanceFormatter
+ */
+class MapboxManeuverApi internal constructor(
+    private val distanceFormatter: DistanceFormatter,
+    private val processor: ManeuverProcessor
+) : ManeuverApi {
+
+    /**
+     * @param formatter contains various instances for use in formatting distance related data
+     * for display in the UI
+     *
+     * @return a [MapboxManeuverApi]
+     */
+    constructor(formatter: DistanceFormatter) : this(formatter, ManeuverProcessor)
+
+    private val mainJobController: JobControl by lazy { ThreadController.getMainScopeAndRootJob() }
+    private var currentManeuverJob: Job? = null
+    private var upcomingManeuverJob: Job? = null
+
+    /**
+     * Given [BannerInstructions] the method goes through the list of [BannerComponents] and returns
+     * a [Maneuver] wrapped inside [ManeuverCallback]
+     * @param bannerInstruction BannerInstructions object representing [BannerInstructions]
+     * @param callback ManeuverCallback contains [ManeuverState.CurrentManeuver]
+     */
+    override fun retrieveManeuver(
+        bannerInstruction: BannerInstructions,
+        callback: ManeuverCallback
+    ) {
+        currentManeuverJob = mainJobController.scope.launch {
```

Comment:
```suggestion
        currentManeuverJob?.cancel()
        currentManeuverJob = mainJobController.scope.launch {
```
Otherwise we have a dangling reference.